[Bug]: arrow builder is nullptr

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master
- Deployment mode(standalone or cluster): cluster
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus-2.0.1
- OS(Ubuntu or CentOS): centos
- CPU/Memory: 32core/196GB
- GPU: 
- Others:
```


### Current Behavior

datnode exits abnormally while user is inserting data

### Expected Behavior

datanode running healthily

### Steps To Reproduce

_No response_

### Milvus Log

```
panic: insertBufferNode flushBufferData failed, err = All attempts results:
attempt #1:[UnexpectedError] arrow builder is nullptr
attempt #2:[UnexpectedError] arrow builder is nullptr
attempt #3:[UnexpectedError] arrow builder is nullptr
attempt #4:[UnexpectedError] arrow builder is nullptr
attempt #5:[UnexpectedError] arrow builder is nullptr


goroutine 17339136 [running]:
github.com/milvus-io/milvus/internal/datanode.(*insertBufferNode).Operate(0xc0007cb590, 0xc002dd4b70, 0x1, 0x1, 0x4256220, 0xc0033a45a0, 0x0)
        /opt/milvus/internal/datanode/flow_graph_insert_buffer_node.go:376 +0x3895
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc003008b00)
        /opt/milvus/internal/util/flowgraph/node.go:102 +0x23b
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
        /opt/milvus/internal/util/flowgraph/node.go:70 +0x70
```

### Anything else?

_No response_

> Unfortunately, I cannot reproduce locally. @BUPTAnderson It would be very nice of you to provide the script or logs to us.

the datanode log:
```
[2022/06/27 18:22:47.140 +08:00] [INFO] [segment_replica.go:625] ["updating segment"] ["Segment ID"=434195724883722241] [numRows=6]
[2022/06/27 18:22:47.140 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3209] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.143 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3209] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.149 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3209] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.152 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3209] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.155 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3209] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.158 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3209] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.166 +08:00] [INFO] [flow_graph_insert_buffer_node.go:249] ["insert seg buffer status"] [No.=0] [segmentID=434195724883722241] ["vchannel name"=by-dev-rootcoord-dml_22_434192070176669697v0] ["buffer size"=3209] ["buffer limit"=8192]
[2022/06/27 18:22:47.340 +08:00] [INFO] [segment_replica.go:625] ["updating segment"] ["Segment ID"=434195724883722241] [numRows=4]
[2022/06/27 18:22:47.340 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3213] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.344 +08:00] [INFO] [segment_replica.go:159] ["update pk range"] [collectionID=434192070176669697] [partitionID=434192073820733441] [segmentID=434195724883722241] [channel=by-dev-rootcoord-dml_22_434192070176669697v0] [num_rows=3213] [minPK=1001711116065925] [maxPK=1002511144059205]
[2022/06/27 18:22:47.344 +08:00] [ERROR] [flow_graph_insert_buffer_node.go:375] ["insertBufferNode flushBufferData failed, err = All attempts results:\nattempt #1:[UnexpectedError] arrow builder is nullptr\nattempt #2:[UnexpectedError] arrow builder is nullptr\nattempt #3:[UnexpectedError] arrow builder is nullptr\nattempt #4:[UnexpectedError] arrow builder is nullptr\nattempt #5:[UnexpectedError] arrow builder is nullptr\n"] [stack="github.com/milvus-io/milvus/internal/datanode.(*insertBufferNode).Operate\n\t/opt/milvus/internal/datanode/flow_graph_insert_buffer_node.go:375\ngithub.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work\n\t/opt/milvus/internal/util/flowgraph/node.go:102"]
```

the `collectionID=434192070176669697` schema:
```
{
  auto_id: False
  description: ci cv finegrain-nextvlad-v2 feedvideo collection
  fields: [{
    name: id
    description: feedid
    type: 5
    is_primary: True
    auto_id: False
  }, {
    name: embedding
    description: feed embedding
    type: 101
    params: {'dim': 512}
  }, {
    name: embed_insert_time
    description: timestamp: second
    type: 5
  }, {
    name: feed_create_time
    description: timestamp: second
    type: 5
  }]
}
```

--- a/internal/core/src/storage/parquet_c.cpp
+++ b/internal/core/src/storage/parquet_c.cpp
@@ -305,6 +305,12 @@ FinishPayloadWriter(CPayloadWriter payloadWriter) {
     st.error_msg = nullptr;
     auto p = reinterpret_cast<wrapper::PayloadWriter*>(payloadWriter);
     if (p->builder == nullptr) {
+        if (p->dimension == wrapper::EMPTY_DIMENSION) {
+            // For FloatVector/BinaryVector datatype, the builder is lazily inited.
+            // Since wrapper::EMPTY_DIMENSION indicates the builder is not inited,
+            // we simply return success here.
+            return st;
+        }
         st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
         st.error_msg = ErrorMsg("arrow builder is nullptr");
         return st;

--- a/internal/core/unittest/test_parquet_c.cpp
+++ b/internal/core/unittest/test_parquet_c.cpp
@@ -248,6 +248,21 @@ TEST(wrapper, binary_vector) {
     ReleasePayloadReader(reader);
 }
 
+TEST(wrapper, binary_vector_empty) {
+    auto payload = NewPayloadWriter(ColumnType::VECTOR_BINARY);
+    auto st = FinishPayloadWriter(payload);
+    ASSERT_EQ(st.error_code, ErrorCode::SUCCESS);
+    auto cb = GetPayloadBufferFromWriter(payload);
+    ASSERT_EQ(cb.length, 0);
+    ASSERT_EQ(cb.data, nullptr);
+    auto nums = GetPayloadLengthFromWriter(payload);
+    ASSERT_EQ(nums, 0);
+    auto reader = NewPayloadReader(ColumnType::VECTOR_BINARY, (uint8_t*)cb.data, cb.length);
+    ASSERT_EQ(reader, nullptr);
+    ReleasePayloadWriter(payload);
+    ReleasePayloadReader(reader);
+}
+
 TEST(wrapper, float_vector) {
     auto payload = NewPayloadWriter(ColumnType::VECTOR_FLOAT);
     float data[] = {1, 2, 3, 4, 5, 6, 7, 8};
@@ -278,6 +293,21 @@ TEST(wrapper, float_vector) {
         ASSERT_EQ(values[i], data[i]);
     }
 
+    ReleasePayloadWriter(payload);
+    ReleasePayloadReader(reader);
+}
+
+TEST(wrapper, float_vector_empty) {
+    auto payload = NewPayloadWriter(ColumnType::VECTOR_FLOAT);
+    auto st = FinishPayloadWriter(payload);
+    ASSERT_EQ(st.error_code, ErrorCode::SUCCESS);
+    auto cb = GetPayloadBufferFromWriter(payload);
+    ASSERT_EQ(cb.length, 0);
+    ASSERT_EQ(cb.data, nullptr);
+    auto nums = GetPayloadLengthFromWriter(payload);
+    ASSERT_EQ(nums, 0);
+    auto reader = NewPayloadReader(ColumnType::VECTOR_FLOAT, (uint8_t*)cb.data, cb.length);
+    ASSERT_EQ(reader, nullptr);
     ReleasePayloadWriter(payload);
     ReleasePayloadReader(reader);
 }

--- a/internal/storage/payload_cgo_test.go
+++ b/internal/storage/payload_cgo_test.go
@@ -583,7 +583,7 @@ func TestPayload_CGO_ReaderandWriter(t *testing.T) {
 		assert.NotNil(t, err)
 
 		err = w.FinishPayloadWriter()
-		assert.NotNil(t, err)
+		assert.Nil(t, err)
 
 		err = w.AddBinaryVectorToPayload([]byte{}, 8)
 		assert.NotNil(t, err)
@@ -604,7 +604,7 @@ func TestPayload_CGO_ReaderandWriter(t *testing.T) {
 		defer w.Close()
 
 		err = w.FinishPayloadWriter()
-		assert.NotNil(t, err)
+		assert.Nil(t, err)
 
 		err = w.AddFloatVectorToPayload([]float32{}, 8)
 		assert.NotNil(t, err)

--- a/internal/storage/payload_test.go
+++ b/internal/storage/payload_test.go
@@ -583,7 +583,7 @@ func TestPayload_ReaderAndWriter(t *testing.T) {
 		assert.NotNil(t, err)
 
 		err = w.FinishPayloadWriter()
-		assert.NotNil(t, err)
+		assert.Nil(t, err)
 
 		err = w.AddBinaryVectorToPayload([]byte{}, 8)
 		assert.NotNil(t, err)
@@ -604,7 +604,7 @@ func TestPayload_ReaderAndWriter(t *testing.T) {
 		defer w.Close()
 
 		err = w.FinishPayloadWriter()
-		assert.NotNil(t, err)
+		assert.Nil(t, err)
 
 		err = w.AddFloatVectorToPayload([]float32{}, 8)
 		assert.NotNil(t, err)

